overlord/state/state.go

```diff
@@ -0,0 +1,158 @@
+	// ExecContext executes a query that doesn't return rows.
+	// For example: an INSERT and UPDATE.
+	ExecContext(context.Context, string, ...interface{}) (sql.Result, error)
+	// QueryContext executes a query that returns rows, typically a SELECT.
+	QueryContext(context.Context, string, ...interface{}) (*sql.Rows, error)
+}
+
+// TxnBuilder allows the building of a series of functions that will be called
+// during a transaction commit. Only upon commit is the transaction constructed
+// and the function called.
+// The functions in the txn builder maybe called multiple times depending on
+// how many retries are employed.
+type TxnBuilder interface {
+	Run(func(context.Context, Txn) error)
+	Commit() error
+}
+
+// Run is a convince function for running one shot transactions, which correctly
+// handles the rollback semantics and retries where available.
+// The run function maybe called multiple times if the transaction is being
+// retried.
+func (s *State) Run(fn func(context.Context, Txn) error) error {
+	txn, err := s.CreateTxn(context.Background())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	txn.Run(fn)
+
+	return txn.Commit()
+}
+
+// CreateTxn creates a transaction builder. The transaction builder accumulates
+// a series of functions that can be executed on a given commit.
+func (s *State) CreateTxn(ctx context.Context) (TxnBuilder, error) {
+	return &txnBuilder{
+		db:  s.DB(),
+		ctx: ctx,
+	}, nil
+}
+
+// txnBuilder creates a type for executing transactions and ensuring rollback
+// symantics are employed.
+type txnBuilder struct {
+	db        *sql.DB
+	ctx       context.Context
+	runnables []func(context.Context, Txn) error
+}
+
+// Context returns the underlying TxnBuilder context.
+func (t *txnBuilder) Context() context.Context {
+	return t.ctx
+}
+
+// Run adds a function to a given transaction context. The transaction
+// isn't committed until the commit method is called.
+// The run function maybe called multiple times if the transaction is being
+// retried.
+func (t *txnBuilder) Run(fn func(context.Context, Txn) error) {
+	t.runnables = append(t.runnables, fn)
+}
+
+// Commit commits the transaction.
+func (t *txnBuilder) Commit() error {
+	return withRetry(func() error {
+		// Ensure that we don't attempt to retry if the context has been
+		// cancelled or errored out.
+		if err := t.ctx.Err(); err != nil {
+			return errors.Trace(err)
+		}
+
+		tx, err := t.db.BeginTx(t.ctx, nil)
+		if err != nil {
+			// Nested transactions are not supported, if we get an error during
+			// the begin transaction phase, attempt to rollback both
+			// transactions, so that they can correctly start again.
+			if tx != nil {
+				_, _ = tx.Exec("ROLLBACK")
+			}
+			return errors.Trace(err)
+		}
+		// TODO (stickupkid): We should wrap this transaction, so it's possible
+		// to abort a transaction within a run function.
+		for _, fn := range t.runnables {
+			if err := fn(t.ctx, tx); err != nil {
+				return errors.Trace(err)
```

Comment:
You need a `tx.Rollback` here to avoid leaking the txn handle

Reply:
Agreed.